core/src/server/rpc_module.rs

```diff
@@ -721,39 +561,73 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 	///
 	/// # Returns
 	///
-	/// An async block which returns `Result<(), SubscriptionCallbackError>` the error is simply
-	/// for a more ergonomic API and is not used (except logged for user-related caused errors).
-	/// By default jsonrpsee doesn't send any special close notification,
-	/// it can be a footgun if one wants to send out a "special notification" to indicate that an error occurred.
+	/// An async block which returns something that implements [`crate::server::IntoSubscriptionResponse`] which
+	/// decides what action to take once the subscription callback returns i.e, whether to sent out another message
+	/// on the subscription stream before closing down it.
+	///
+	/// This is implemented for `Result<T, E>` and `Option<T>`.
+	///
+	/// It's recommended to use `Result` if you want to propagate the error as special error notification
```

Comment:
I haven't checked yet, but if the subscription returns an error before `pending.accept()` does that also come back (but as an error to the subscription request rather than an error notification). Maybe the text should be more general and just say something like "Use `Result` if you want the error to be sent back to the user. Use `Option` if you'd like to avoid sending any errors back to the user."?

Reply:
Ah nevermind; it's a different thing and the user has to `reject()` with an error, which makes sense, so ignore this :)

Reply:
This applies only to subscription notifications i.e, subscription messages and before that the subscription has been accepted we don't allow any messages to be sent. So those are ignored unless the `reject` is called explicitly. 

However, it would be nice to incorporate that in the callback and let remove `reject API`

TLDR: Returning stuff before `accept` has no effect.

Reply:
So, I want to keep it out of this PR but indeed this is hairy stuff to understand.

Reply:
Yeah makes sense; I can imagine that if the user returns an IntoResponse thing before calling `pending.accept()`, that maybe it could return the error they gave back (if a `Result` type was used) or just return some default error. And then maybe keep `pending.reject(explicit_error)` around so that even if the user wants to return `Option` from the subscription, they can still send a non-default error back here if they want.

But yeah, definitely can be a separate PR if wanted :)